### ISSUE TITLE
Bug during XML deserialization : "com.fasterxml.jackson.databind.exc.MismatchedInputException: Cannot deserialize instance of java.lang.Double out of START_OBJECT token"

### ISSUE THREAD

**Issue by JulienHoueix** [FasterXML/jackson-dataformat-xml#301]
Hello,

I encountered a bug during XML deserialization with Jackson 2.9.6

If I have a parent class that :
- contains a list of objects with a specific name
- and also contains another list of different objects which contain a value with that same name

I get the following error when I try to serialize the XML :

_com.fasterxml.jackson.databind.exc.MismatchedInputException: Cannot deserialize instance of `java.lang.Double` out of START_OBJECT token
 at [Source: (StringReader); line: 1, column: 44] (through reference chain: ch.hcuge.deja.drug.client.Parent["CHILDB"]->java.util.ArrayList[0]->ch.hcuge.deja.drug.client.ChildB["MY_PROPERTY"])_

**But I have this error only if the parent object contains an attribute!**

I have a test case that shows the problem :
```
import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.dataformat.xml.XmlMapper;
import com.fasterxml.jackson.dataformat.xml.annotation.JacksonXmlElementWrapper;
import com.fasterxml.jackson.dataformat.xml.annotation.JacksonXmlProperty;
import org.junit.Test;

import java.io.StringWriter;
import java.util.ArrayList;
import java.util.List;

public class TestBug {

    // Succeeds
    @Test
    public void testWithoutAttribute() throws Exception {
        StringWriter stringWriter = new StringWriter();
        stringWriter.append("" +
                "<PARENT>" +
                "  <CHILDB>" +
                "    <MY_PROPERTY>12.34</MY_PROPERTY>" +
                "  </CHILDB>" +
                "</PARENT>");
        XmlMapper xmlMapper = new XmlMapper();
        Parent deserializedParent = xmlMapper.readValue(stringWriter.toString(), Parent.class);
    }

    // Fails!
    @Test
    public void testWithAttribute() throws Exception {
        StringWriter stringWriter = new StringWriter();
        stringWriter.append("" +
                "<PARENT>" +
                "  <CHILDB MY_ATTR=\"TEST_VALUE\">" +
                "    <MY_PROPERTY>12.34</MY_PROPERTY>" +
                "  </CHILDB>" +
                "</PARENT>");
        XmlMapper xmlMapper = new XmlMapper();
        Parent deserializedParent = xmlMapper.readValue(stringWriter.toString(), Parent.class);
    }

}

class Root {

    @JacksonXmlProperty(localName = "PARENT")
    public Parent parent;

}

@JsonIgnoreProperties(ignoreUnknown = true)
class Parent {

    @JacksonXmlProperty(localName = "MY_ATTR", isAttribute = true)
    public String myAttribute;

    @JacksonXmlElementWrapper(useWrapping = false)
    @JacksonXmlProperty(localName = "MY_PROPERTY")
    public List<ChildA> childrenA = new ArrayList<>();

    @JacksonXmlElementWrapper(useWrapping = false)
    @JacksonXmlProperty(localName = "CHILDB")
    public List<ChildB> childrenB = new ArrayList<>();

}

class ChildA {

}

@JsonIgnoreProperties(ignoreUnknown = true)
@JsonInclude(JsonInclude.Include.NON_NULL)
class ChildB {

    @JacksonXmlProperty(localName = "MY_PROPERTY")
    public Double value;

}
```

**Comment by JulienHoueix** [comment]
I looked at the Jackson code.
It seems that the problem is related to JacksonXmlElementWrapper annotations with "useWrapping = false".

In class "com.fasterxml.jackson.dataformat.xml.deser.FromXmlParser", there is the following code in the method "nextToken()" : 

>
            // Ok: virtual wrapping can be done by simply repeating current START_ELEMENT.
            // Couple of ways to do it; but start by making _xmlTokens replay the thing...
            if (_namesToWrap != null && _namesToWrap.contains(name)) {
                _xmlTokens.repeatStartElement();
            }

where "_namesToWrap" contains the property lists that have "useWrapping = false".

The problem is that when we are parsing "<MY_PROPERTY>12.34</MY_PROPERTY>" of "CHILDB", the list "_namesToWrap" contains "MY_PROPERTY" and "CHILD_B", so it considers that "MY_PROPERTY" should be an unwrapped list instead of a Double value.

I think the problem is that the property "_namesToWrap" is not up to date in that part of the code. It works if I replace the code with the following :
>
            // Ok: virtual wrapping can be done by simply repeating current START_ELEMENT.
            // Couple of ways to do it; but start by making _xmlTokens replay the thing...
            _namesToWrap = _parsingContext.getNamesToWrap();
            if (_namesToWrap != null && _namesToWrap.contains(name)) {
                _xmlTokens.repeatStartElement();
            }

PS : I was willing to do a PR, but there are several tests that do not pass (EnumIssue9Test for example) when I retrieved the code, so I don't really know if it creates regressions.

**Comment by mkamalas** [comment]
@JulienHoueix  - Please advise if you found a workaround for this issue.

**Comment by JulienHoueix** [comment]
> @JulienHoueix - Please advise if you found a workaround for this issue.

No

**Comment by DonaldMaparura** [comment]
you can use json and include your attributes in an object

**Comment by cowtowncoder** [comment]
Hoping to look more into this today, but one quick note: whenever troubleshooting I think it is very important to remove these entries

```
@JsonIgnoreProperties(ignoreUnknown = true)
```

since this very often hides real problems and makes it more difficult to see the issue.
Same for global configuration.

**Comment by cowtowncoder** [comment]
Got

```com.fasterxml.jackson.databind.exc.UnrecognizedPropertyException: Unrecognized field "MY_ATTR" (class com.fasterxml.jackson.dataformat.xml.lists.List301DeserTest$ChildB), not marked as ignorable (one known property: "MY_PROPERTY"])
 at [Source: (StringReader); line: 1, column: 40] (through reference chain: 
```

after removing `@JsonIgnoreProperties`, so I can reproduce the problem with 2.11.0.

**Comment by cowtowncoder** [comment]
Oh. Actually... putting back `@JsonIgnoreProperties(ignoreUnknown = true)` will make test pass. And looking at structure, the real problem seems to be that 

```
        @JacksonXmlProperty(localName = "MY_ATTR", isAttribute = true)
        public String myAttribute;
```

is in `Parent` class, but XML content 

```xml
<PARENT>
  <CHILDB MY_ATTR='TEST_VALUE'>
    <MY_PROPERTY>12.34</MY_PROPERTY>
  </CHILDB>
</PARENT>
```

expects it for `ChildB`.

So I think this is fixed for 2.11.1 (to be released) at least, possibly earlier. Not sure which of fixes since 2.9.6 it would be (there are a few), but will mark fixed for 2.11.1
